File: packages/python/minions_openclaw/instance_manager.py
"""Instance manager - Python equivalent of TypeScript InstanceManager."""
from __future__ import annotations
import json
from pathlib import Path
from typing import List, Optional, Dict, Any

from ._minions_stub import Minion, Relation, create_minion, soft_delete, generate_id, now
from .types import openclaw_instance_type

DATA_DIR = Path.home() / '.openclaw-manager'
DATA_FILE = DATA_DIR / 'data.json'


def _read_storage() -> Dict[str, Any]:
    DATA_DIR.mkdir(parents=True, exist_ok=True)
    try:
        return json.loads(DATA_FILE.read_text())
    except (FileNotFoundError, json.JSONDecodeError):
        return {'minions': [], 'relations': []}


def _write_storage(data: Dict[str, Any]) -> None:
    DATA_DIR.mkdir(parents=True, exist_ok=True)
    DATA_FILE.write_text(json.dumps(data, indent=2))


def _minion_from_dict(d: Dict[str, Any]) -> Minion:
    return Minion(
        id=d['id'],
        title=d['title'],
        minion_type_id=d['minionTypeId'],
        fields=d.get('fields', {}),
        created_at=d.get('createdAt', now()),
        updated_at=d.get('updatedAt', now()),
        tags=d.get('tags', []),
        status=d.get('status', 'active'),
        priority=d.get('priority', 'medium'),
        description=d.get('description', ''),
        deleted_at=d.get('deletedAt'),
    )


def _minion_to_dict(m: Minion) -> Dict[str, Any]:
    d: Dict[str, Any] = {
        'id': m.id,
        'title': m.title,
        'minionTypeId': m.minion_type_id,
        'fields': m.fields,
        'createdAt': m.created_at,
        'updatedAt': m.updated_at,
        'tags': m.tags,
        'status': m.status,
        'priority': m.priority,
        'description': m.description,
    }
    if m.deleted_at:
        d['deletedAt'] = m.deleted_at
    return d


class InstanceManager:
    def register(self, name: str, url: str, token: Optional[str] = None) -> Minion:
        fields: Dict[str, Any] = {'url': url, 'status': 'registered'}
        if token:
            fields['token'] = token
        minion, validation = create_minion(name, openclaw_instance_type, fields)
        if not validation.valid:
            raise ValueError(f"Validation failed: {[e.message for e in validation.errors]}")
        storage = _read_storage()
        storage['minions'].append(_minion_to_dict(minion))
        _write_storage(storage)
        return minion

    def list(self) -> List[Minion]:
        storage = _read_storage()
        return [
            _minion_from_dict(m) for m in storage['minions']
            if m.get('minionTypeId') == openclaw_instance_type.id
            and not m.get('deletedAt')
        ]

    def get_by_id(self, id: str) -> Optional[Minion]:
        storage = _read_storage()
        for m in storage['minions']:
            if m.get('id') == id and not m.get('deletedAt'):
                return _minion_from_dict(m)
        return None

    def remove(self, id: str) -> None:
        storage = _read_storage()
        for i, m in enumerate(storage['minions']):
            if m.get('id') == id:
                minion = _minion_from_dict(m)
                deleted = soft_delete(minion)
                storage['minions'][i] = _minion_to_dict(deleted)
                _write_storage(storage)
                return
        raise ValueError(f"Instance {id} not found")

File: packages/core/src/InstanceManager.ts
import { createMinion, updateMinion, softDelete, generateId } from 'minions-sdk';
import type { Minion, Relation } from 'minions-sdk';
import { openclawInstanceType } from './types.js';
import { GatewayClient } from './GatewayClient.js';
import { promises as fs } from 'fs';
import { homedir } from 'os';
import { join } from 'path';

const DATA_DIR = join(homedir(), '.openclaw-manager');
const DATA_FILE = join(DATA_DIR, 'data.json');

interface StorageData {
  minions: Minion[];
  relations: Relation[];
}

async function readStorage(): Promise<StorageData> {
  try {
    await fs.mkdir(DATA_DIR, { recursive: true });
    const raw = await fs.readFile(DATA_FILE, 'utf-8');
    return JSON.parse(raw) as StorageData;
  } catch {
    return { minions: [], relations: [] };
  }
}

async function writeStorage(data: StorageData): Promise<void> {
  await fs.mkdir(DATA_DIR, { recursive: true });
  await fs.writeFile(DATA_FILE, JSON.stringify(data, null, 2), 'utf-8');
}

export class InstanceManager {
  private clients = new Map<string, GatewayClient>();

  async register(name: string, url: string, token?: string): Promise<Minion> {
    const { minion, validation } = createMinion(
      {
        title: name,
        fields: {
          url,
          ...(token ? { token } : {}),
          status: 'registered',
        },
        tags: [],
      },
      openclawInstanceType
    );
    if (!validation.valid) {
      throw new Error(`Validation failed: ${validation.errors.map(e => e.message).join(', ')}`);
    }
    const storage = await readStorage();
    storage.minions.push(minion);
    await writeStorage(storage);
    return minion;
  }

  async connect(instanceId: string): Promise<GatewayClient> {
    const instance = await this.getById(instanceId);
    if (!instance) throw new Error(`Instance ${instanceId} not found`);
    const url = instance.fields['url'] as string;
    const token = instance.fields['token'] as string | undefined;
    const devicePrivateKey = instance.fields['devicePrivateKey'] as string | undefined;
    const client = new GatewayClient(url, token, devicePrivateKey);
    this.clients.set(instanceId, client);
    await client.openConnection();
    return client;
  }

  async ping(instanceId: string): Promise<number> {
    const start = Date.now();
    const client = await this.connect(instanceId);
    const latency = Date.now() - start;
    await client.close();
    this.clients.delete(instanceId);
    const storage = await readStorage();
    const idx = storage.minions.findIndex(m => m.id === instanceId);
    if (idx !== -1) {
      const { minion } = updateMinion(
        storage.minions[idx],
        {
          fields: {
            ...storage.minions[idx].fields,
            lastPingAt: new Date().toISOString(),
            lastPingLatencyMs: latency,
            status: 'online',
          },
        },
        openclawInstanceType
      );
      storage.minions[idx] = minion;
      await writeStorage(storage);
    }
    return latency;
  }

  async disconnect(instanceId: string): Promise<void> {
    const client = this.clients.get(instanceId);
    if (client) {
      await client.close();
      this.clients.delete(instanceId);
    }
  }

  async list(): Promise<Minion[]> {
    const storage = await readStorage();
    return storage.minions.filter(
      m => m.minionTypeId === openclawInstanceType.id && (m.deletedAt === undefined || m.deletedAt === null)
    );
  }

  async getById(id: string): Promise<Minion | undefined> {
    const storage = await readStorage();
    return storage.minions.find(m => m.id === id && (m.deletedAt === undefined || m.deletedAt === null));
  }

  async remove(id: string): Promise<void> {
    const storage = await readStorage();
    const idx = storage.minions.findIndex(m => m.id === id);
    if (idx === -1) throw new Error(`Instance ${id} not found`);
    storage.minions[idx] = softDelete(storage.minions[idx]);
    await writeStorage(storage);
  }
}
